### PR TITLE
WooCommerce Deposits Multi Currency compatibility

### LINKED ISSUE #1902: Customer facing prices compatibility with WooCommerce Deposits
Follow up from #1769

We need to ensure compatibility with Customer multi-currency prices and WooCommerce Deposits.

### PR DESCRIPTION
**Pull request by tpaksu**
Fixes #1902

#### Changes proposed in this Pull Request

This PR adds compatibility between WooCommerce Payments Multi-currency and WooCommerce Deposits Plugin. With this PR, the customers can buy products with splittable payment options with any currency the merchant enabled for their site. Before this PR:
- The deposit amount wasn't calculated in the correct currency
- The future orders were created with the store currency, not with the original order currency.

##### Product with a single deposit option

- **Preparation Steps**

  * Install the WooCommerce Deposits plugin and activate it.
  * Go to one of the products' admin page and see the "Deposits" tab on the general product settings section.
  * Enable deposits for that product 
  * Select "Deposit Type" as "Percentage" or "Fixed Amount"
  * Enter a fixed amount or percentage relative to your selection on the previous step
  * Update the product by clicking the "Update" button
  
- **Testing**
  * Go to the shop page and sell the item
  * Check the prices on the cart dropdown and the cart page
  * Go to checkout and check the prices
  * Make the purchase and check the prices on the confirmation page
  * Go to WooCommerce > Orders page, select the order you just created
  * Check the prices there, and also you should see a "Invoice remaining balance" button
  * Click it. Check the prices and the currency on the new order created for the remaining amount.
  

##### Product with a payment plan type deposit option

- **Preparation Steps**

  * Create a payment plan on "Admin > Products > Payment Plans" page, for testing the prices easily, you can select percentages like 10%, 50% etc. 
  * After you've created the payment plan, go to one of the products' admin page and see the "Deposits" tab on the general product settings section.
  * Enable deposits for that product 
  * Select "Deposit Type" as "Payment Plan"
  * Select the payment plan you just created.
  * Update the product by clicking the "Update" button
  
- **Testing**
  * Go to the shop page and sell the item
  * Check the prices on the cart dropdown and the cart page
  * Go to checkout and check the prices
  * Make the purchase and check the prices on the confirmation page
  * Go to WooCommerce > Orders page, select the order you just created
  * Check the prices there, and also you should see a "View Scheduled Payments" button
  * Click it. Check the scheduled orders for the remaining payment steps.

**Extra testing:** Create a discount coupon and redo the tests with a coupon entered in the cart page

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

**Review by jessepearson — CHANGES_REQUESTED** [review]
When viewing a deposit product, the deposit amount is not converted
Markup 2021-11-05 at 09 47 11

I believe there's a rounding issue that is going to cause problems with zero decimal currencies. If your payment plan is 50% up front, and then 20%, 20%, 10%, with an initial cost of $100 USD,  then with JPY at `113.65175`, the totals do not add up or round correctly.

The below is the original 50% order with the full amount not rounded to the nearest decimal. After that is payment 2 (which is the same as 3), and payment 4, where the totals do not add up due to rounding.
Markup 2021-11-05 at 10 14 28
Markup 2021-11-05 at 10 15 09
Markup 2021-11-05 at 10 15 40

We've also talked in Slack about the tests. I believe they are going to be hard, but we should have them if we can do them.

**Comment by tpaksu** [comment]
> I believe there's a rounding issue that is going to cause problems with zero decimal currencies. If your payment plan is 50% up front, and then 20%, 20%, 10%, with an initial cost of $100 USD, then with JPY at 113.65175, the totals do not add up or round correctly.

Are you talking about a rounding problem? Because 6251+2501+2501+1250 adds up to 12503, which is the rounded value of 12502,6. And The sum of the subtotals equal to 11366.

**Comment by jessepearson** [comment]
> Are you talking about a rounding problem? Because 6251+2501+2501+1250 adds up to 12503, which is the rounded value of 12502,6. And The sum of the subtotals equal to 11366.

They may add up in the end, but if the orders don't show the correct subtotal and tax amounts, merchants and their customers will notice and it can lead to a bad user experience. 

I disabled MC, set my currency to JPY, and set the store's decimals to `0`, then duplicated my product and set the initial price to `11366` so that I could try to replicate the result of what happens with MC. This is what I see:

Markup 2021-11-05 at 10 48 33
Markup 2021-11-05 at 10 48 58
Markup 2021-11-05 at 10 49 20

I think the issue might be related to the decimals used for the order. If the store's decimals are 2 instead of 0, the rounding changes. Deposits might have something that checks the decimals that we can modify.

File: includes/multi-currency/Compatibility/WooCommerceDeposits.php [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+			add_filter( 'woocommerce_get_cart_contents', [ $this, 'modify_cart_item_deposit_amounts' ] );
+			add_filter( 'woocommerce_product_get__wc_deposit_amount', [ $this, 'modify_cart_item_deposit_amount_meta' ], 10, 2 );
+			add_filter( MultiCurrency::FILTER_PREFIX . 'should_convert_product_price', [ $this, 'maybe_convert_product_prices_for_deposits' ], 10, 2 );
+		}
+	}
+
+	/**
+	 * Converts the currency for deposit amounts of each cart item, only applies if deposits are enabled on the product.
+	 *
+	 * @param array $cart_contents The current tax definitions.
```
Comment (jessepearson) [inline includes/multi-currency/Compatibility/WooCommerceDeposits.php:63]: Linter bot: Docblock needs update 😀

File: includes/multi-currency/Compatibility/WooCommerceDeposits.php [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+			}
+		}
+
+		return $cart_contents;
+	}
+
+	/**
+	 * Converts the currency for deposit amounts of each cart item, only applies if deposits are enabled on the product.
+	 *
+	 * @param array $cart_contents The current tax definitions.
```
Comment (jessepearson) [inline includes/multi-currency/Compatibility/WooCommerceDeposits.php:81]: Linter bot: Docblock needs update 🤖

File: includes/multi-currency/Compatibility/WooCommerceDeposits.php [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+	 * Defines if the product prices need to be converted when calculating totals,
+	 * if the product's deposit type is a payment plan, then it shouldn't convert it.
+	 *
+	 * @param   bool        $result   The previous flag for converting the price.
+	 * @param   \WC_Product $product  The product to check for.
+	 *
+	 * @return  bool                        Whether the price should be converted or not.
+	 */
+	public function maybe_convert_product_prices_for_deposits( $result, $product ) {
+		if ( class_exists( 'WC_Deposits_Product_Manager' )
```
Comment (jessepearson) [inline includes/multi-currency/Compatibility/WooCommerceDeposits.php:103]: I think it's safe _not_ to test to see if `WC_Deposits_Product_Manager` exists like in `modify_cart_item_deposit_amount_meta` due to this should only be called if Deposits is installed and active. I don't think the Deposits extension will have a large enough change to remove that class or method.

**Review by jessepearson — APPROVED** [review]
Pre-approving this pending the docblock and `class_exists` changes mentioned.

Also, I think we can leave the rounding issue for now, since Deposits itself doesn't handle zero decimal currencies correctly, as we discussed in Slack p1636120522121700-slack-C01B8KNUYSW

**Comment by jessepearson** [comment]
@tpaksu There was one other thing I missed adding in my review above, my error. The order payment page the customer is taken to in order pay for the remainder of an order or for one of the payment plan orders should have the currency switcher widget hidden. Having it there allows them to switch the currency, which actually causes an error because it clears out the order variables in the url.

**Comment by tpaksu** [comment]
> @tpaksu There was one other thing I missed adding in my review above, my error. The order payment page the customer is taken to in order pay for the remainder of an order or for one of the payment plan orders should have the currency switcher widget hidden. Having it there allows them to switch the currency, which actually causes an error because it clears out the order variables in the url.

That sounds like something hard to pursue, but definitely needs to be done. Thanks for thinking this!

File: includes/multi-currency/Compatibility/WooCommerceDeposits.php (hunk 3 above)
Comment (tpaksu) [reply includes/multi-currency/Compatibility/WooCommerceDeposits.php:104]: Yes, psalm also agrees that :)

**Comment by tpaksu** [comment]
> Having it there allows them to switch the currency, which actually causes an error because it clears out the order variables in the url.

I've tested it with the order for a remaining amount, clicked the "Customer Payment Page" link on the admin side, and also tried on the "My account" page on the client side, changed the currency via the currency switcher, the checkout currency didn't change and I was able to pay with the order currency. Am I missing something here @jessepearson ? Can you tell me the steps to reproduce the error?

**Comment by jessepearson** [comment]
@tpaksu I just checked, and it's the block widget vs the legacy widget. The block widget clears out the parameters in the url, which causes an error when the page refreshes. It looks like the issue was fixed once, but we will have to look into it again. Created new issue here: https://github.com/Automattic/woocommerce-payments/issues/3317

**Pull request merged**